api/wasm/cpp/proxy_wasm_api.h

```diff
@@ -522,6 +523,32 @@ inline WasmResult getPluginDirection(PluginDirection *direction_ptr) {
   PROXY_EXPRESSION_GET_UINT32("plugin.direction", reinterpret_cast<uint32_t*>(direction_ptr));
 }
 
+// Generic selector
+inline absl::optional<WasmDataPtr> getSelectorExpression(std::initializer_list<absl::string_view> parts) {
+  size_t size = 0;
+  for (auto part: parts) {
+    size += part.size() + 1; // null terminated string value
+  }
+
+  char* buffer = static_cast<char*>(::malloc(size));
+  char* b = buffer;
+
+  for (auto part : parts) {
+    memcpy(b, part.data(), part.size());
+    b += part.size();
+    *b++ = 0;
```

Comment:
Yes, this encoding is only meant for symbols/strings. We should use a real marshalling protocol for binary data.

Reply:
Ideally, we'd have one way for passing an array/list of any type of data.

Reply:
Why? You're going to lose a lot of efficiency if you don't share prior knowledge about the type of data. Genericity is not free.

Reply:
My point was that we could use serialization that works for both, that is, instead of using:
```
<string1.data>\0<string2.data>\0<string3.data>\0
Hello\0world\0!\0
```
we could be using:
```
<string1.len><string1.data><string2.len><string2.data><string3.len><string3.data>
\5Hello\5world\1!
```
which works for both strings and binary values.

I don't think you'd lose any efficiency in doing that (other than a few more bytes for the length).

But in order to unblock this, I'm fine with merging this as-is and moving the discussion to the ABI doc, once I ship it.

Reply:
That's fair. Let's address this later. My concern is that it should support streaming/yielding. Other than that, any scheme is fine.